### DISCUSSION TITLE
How to change the content of a DataTemplate depending on its input item?

### DISCUSSION THREAD

**Discussion by bmscodespace** [AvaloniaUI/Avalonia#16429] [Q&A] [answered]
Hi my fellow Avalonians,

I am having the following problem:
Suppose we want something like this:
```...
 <ItemsRepeater ItemsSource="...">
          <ItemsRepeater.ItemTemplate>
                      <DataTemplate DataType="some complicated Type X">
                                     <some view depending on say a field of our object of type X/>
                      </DataTemplate>
         </ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```  
Changing  the content of our DataTemplate by changing the visibility of its elements does not always work, since there might be controls who are invisible but nonetheless there and they could get input values from our object of type X which they can not deal with. F.e. our object of type X could have an image location as one of its fields and we have two controls (sitting inside of our template), one for displaying gifs and another for displaying all other forms of images. The one that is invisible might not be happy with a particular location string, ending f.e. with "jpeg"  but expecting always "gif". This is just a simple example. 
Therefore the answer to this might be helpful ;).

Thank you very much for any suggestion ...

**Comment by tobyfirth** [comment]
If you can split the contents of your type into multiple type using inheritance or just separate types then you can do something like this.

Rather than using `ItemsRepeater.ItemTemplate` use `ItemsRepeater.DataTemplates`. The code would look like this

```
 <ItemsRepeater ItemsSource="...">
          <ItemsRepeater.DataTemplates>
                      <DataTemplate DataType="Type X">
                                     <some view depending on say a field of our object of type X/>
                      </DataTemplate>
                      <DataTemplate DataType="Type Y">
                                     <some view depending on say a field of our object of type Y/>
                      </DataTemplate>
         </ItemsRepeater.DataTemplates>
</ItemsRepeater>
```

**Reply by bmscodespace** [reply]
Thank you for your answer. Hm... this would be a possibility... Would be cool if there was something like ngIf in Angular ..

**Reply by tobyfirth** [reply]
There are another couple of possibilities.
You could go into the code behind and write something that constructs the group of controls at the appropriate moment based on the state of the object. Or you might be able to do a similar thing using [Avalonia XAML Behaviors](https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors), which would allow you to at least make the solution more reusable and more controllable from the xaml.
I would still recommend using multiple data templates if possible as the easiest way to get something like the behaviour you want.

**Reply by thevortexcloud** [reply, marked as answer]
Or you can just make a custom data template. There is full docs for it here:

https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/IDataTemplateSample

**Reply by tobyfirth** [reply]
I hadn't thought of that one. Probably the best solution in this case.

**Reply by bmscodespace** [reply]
I must say I am more and more convinced of the DataTemplate-usage. Thank you to you all for your comments ...

**Reply by thevortexcloud** [reply]
It definitely takes a bit to get used to, but it's pretty flexible once you get a hang of it. Anyway, you are welcome!
